condition/types.go

```diff
@@ -0,0 +1,193 @@
+package condition
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/google/uuid"
+	"go.hollow.sh/toolbox/events"
+
+	"golang.org/x/exp/slices"
+)
+
+// Kind holds the value for the Condition Kind field.
+type Kind string
+
+const (
+	ServerResourceType    string = "servers"
+	ConditionResourceType string = "condition"
+
+	ConditionCreateEvent events.EventType = "create"
+	ConditionUpdateEvent events.EventType = "update"
+
+	// ConditionStructVersion identifies the condition struct revision
+	ConditionStructVersion string = "1"
+)
+
+// State is the state value of a Condition
+type State string
+
+// Defines holds the value for the Condition State field.
+const (
+	Pending   State = "pending"
+	Active    State = "active"
+	Failed    State = "failed"
+	Succeeded State = "succeeded"
+)
+
+// States returns available condition states.
+func States() []State {
+	return []State{
+		Active,
+		Pending,
+		Failed,
+		Succeeded,
+	}
+}
+
+// Transition valid returns a bool value if the state transition is allowed.
+func (current State) TransitionValid(next State) bool {
+	switch {
+	// Pending state can stay in Pending or transition to Active or Failed or Succeeded
+	case current == Pending && slices.Contains([]State{Pending, Active, Failed, Succeeded}, next):
+		return true
+	// Active state can stay in Active or transition to Failed or Succeeded
+	case current == Active && slices.Contains([]State{Active, Failed, Succeeded}, next):
+		return true
+	default:
+		return false
+	}
+}
+
+// StateValid validates the State.
+func StateIsValid(s State) bool {
+	return slices.Contains(States(), s)
+}
+
+// StateComplete returns true when the given state is considered to be final.
+func StateIsComplete(s State) bool {
+	return slices.Contains([]State{Failed, Succeeded}, s)
+}
+
+// Parameters is an interface for Condition Parameter types
+type Parameters interface {
+	Validate() error
+}
+
+// Definition holds the default parameters for a Condition.
+type Definition struct {
+	Kind                  Kind `mapstructure:"kind"`
+	OnSuccessCondition    Kind `mapstructure:"onSuccessCondition"`
```

Comment:
Since a Condition could have two out comes, it made sense to have a `OnSuccessCondition` / `OnFailureCondition`, would we want to call it a `Task`? since its only a Task after a worker has picked it - I guess we need to settle on the terminology here

Reply:
I also wonder if that needs to be strictly configurable. To give you an idea of what I mean, consider `FirmwareInstall`. Using this definition, you could configure it with a follow-on action, or not and that's a run-time decision which can be changed _ad hoc_. I feel like we're trying to configure steps that should be either coded into the task (i.e. have ConditionOrc *always* perform inventory after a flash), or be a parameter (conditionally perform inventory after a flash). It feels like making this a parameter removes some debuggability by making it harder to understand the relationship between a given firmware install and any follow-on activities.

Reply:
Yeah, you're right - we're tying a set of actions together that must be done, 
do you have thoughts on how we could we have them executed in succession - without the client specifying it themselves or hardcoding it into the orchestrator?

Reply:
Resolving as this is a bigger discussion and tangential to this PR.